### ISSUE TITLE
if Promise is returned element, block further cell evaluation on further Promise resolution / rejection / error

### ISSUE THREAD

**Issue by hamilton** [iodide-project/iodide#597]

**Comment by hamilton** [comment]
After further thinking, I don't think this will necessarily solve the Pyodide issue of loading packages, and in fact will create new problems elsewhere.

- (@mdboom this is relevant for pyodide) for other language integration that requires something async (via WebAssembly) this will not solve our fundamental issues of blocking further cell execution upon resolution. It might make  something like this work, if `loadPackage` is a Promise:

```python
pyodide.loadPackage('pandas')
```

But this will not:

```python
import pyodide
pyodide.loadPackage("pandas")
import pandas as pd

df = pd.read_csv(pyodide.open_url("music-reviews.csv"))
df.describe(include='all')
```

- there are compelling reasons _not_ to force a `fetch` or some other Promise to block further cell evaluation upon resolution, and by forcing this on users we are limiting the utility of Promises / iodide generally.
- if users need a way of pulling in their data sets / external data via a `fetch`-like interface (but don't really want to learn how to use the `iodide.evalQueue` api or Promises generally) then perhaps we should actually provide a simple stdlib async data pull that does this as a convenience.

So I think to solve the pyodide issue we should think about whether there is something we can call or use to make this work. If pyodide have access to the iodide API, then perhaps we can wrap `loadPackage` to also use `iodide.evalQueue.requireExlicitContinuation()` and `iodide.evalQueue.continue()`.

**Comment by mdboom** [comment]
To get your second example to work, one would have to do something like:

```python
def do_rest():
    import pandas as pd

    df = pd.read_csv(pyodide.open_url("music-reviews.csv"))
    df.describe(include='all')

import pyodide
pyodide.loadPackage("pandas").then(do_rest)
```

Which obviously is worse than the Javascript equivalent since Python doesn't have good anonymous functions etc.  So I sort of saw the "block on returned promises" as syntactic sugar around this problem.  Note also, that `loadPackage` is a Javascript function, not a Python one, since I haven't yet figured out a way to integrate async Javascript with Python.

But I see your point that we are forcing one way of thinking about promises on the user that may not be the right one in all cases.  I think I agree with @bcolloran that it's reasonable that block-on-promise is the default.  However, it also breaks the concept that, at least logically, if you took all of the code cells and put then in a source file and ran it, you would have the same result -- i.e. there's implicit "wait and then" in there.

I don't mind hiding the complexity of the blocking inside `loadPackage`: in that case blocking always makes sense and the details of the evalQueue would be hidden from the user.

**Comment by hamilton** [comment]
Yep, that `loadPackage` returns a Promise was the dead giveaway :)

I'm actually just not sure this is the right way to do this, fundamentally. even in the case where `loadPackage` blocks further cell evaluation, we will still have the problem of (new) users probably doing something like this and expecting it to work:

```python
pyodide.loadPackage("pandas")
import pandas as pd
```

I'm actually wondering in this case if we should consider the environment loading within the language plugin, but as an option (like an extra key-value pair like `packages: ["pandas", "numpy", ...]`. That way the environment is defined python-side. We'd retain the modularity, and the async stuff will already be appropriately chained, cell-wise. The downside with this approach is it is less easy to say to yourself "oh, I need library x here it turns out" - you'd have to go back to the language plugin cell and add it. I can, however, see a way that we can do that & prevent reloading the whole language stack as well - just the modules that weren't loaded. Perhaps in theory if the language is already loaded, and a package is added, then it adds the package and blocks further cell evaluation.

Just thinking of ways to keep python users from having to deal with any async hiccups out the gate. We should probably file another issue around this!

**Comment by bcolloran** [comment]
this is not something i know a lot about, so i'm trying to understand the issue. @hamilton, re:
"there are compelling reasons not to force a fetch or some other Promise to block further cell evaluation upon resolution, and by forcing this on users we are limiting the utility of Promises / iodide generally."
can you say more about the compelling reasons and the limitation on the utility? Esp wrt limiting the utility?

like, if we did wait by default, i could see it opening up potentially nice patterns like:
https://iodide-project.github.io/master/?gist=bcolloran/2e104b3fc28f68454a23626e11bc8678
(the "load all you data async, then do stuff with it when it's all ready" pattern) 
which would run-all out of the box if we waited on promise resolution, but which ATM requires using the evalQueue API. I am very willing to believe that there are problems i'm not aware of, but i like this pattern -- it seems to _increase_ the utility of Promises / iodide generally.

regarding @mdboom's comment re: running code in a source file -- that is a nice to have, but not a requirement -- when we build a server side node iodide-runner, the input will be iodide NBs (JSMD), and we'll have some additional control, so i think it's legit for us to have a bit of leeway if it improves the user experience.

but also: i think that proposal for having something like an "imports" or "modules"  or "installModules" section of the language plugin definition seems pretty reasonable / semantically clear from a user point of view. It somehow feel like a cleaner mental model (i'm just wondering if the extra "pyodide.loadPackage" call before importing will look/feel weird to python users). but i don't have strng opinions on this, happy to follow your lead @mdboom

**Comment by hamilton** [comment]
That's a nice pattern for sure, but it fundamentally breaks normal JS reasoning about how promises / eval works for really only a little bit of gain on the initial data exploration side. Maybe that's enough to warrant a change - after all, we need to make data importation dead-simple and right now it isn't. But if we are worried about non-JS-heavy users trying to figure out how to write JS code and block cell evaluation, then I really think we should provide even easier convenience functions so they don't have to learn how Promises / async stuff works before being able to do real work. We can't wish away async from JS generally, but we can make it more convenient for all the different types of workflows we are envisioning.

With that in mind, here are a few ideas based on thinking + other convos we've had:

- yes, let's block further cell evaluation on the last Promise. But if I'm switching from a data exploration to a presentation of some kind, I'm going to need to have a way of unblocking cells that maybe have a Promise as the last call so things run kind of async-like when I'm ready for them to. 
- __`iodide.fetch`__ Most analysts are going to primarily encounter Promises when they load data, so perhaps we can easily provide a custom fetch function that exactly works the same way as `fetch` (`iodide.fetch` or something like that) but also prevents further cell evaluation upon Promise resolution / rejection. We can even have it so that every time `iodide.fetch` gets called in a cell, we can wait for all the various instances of that call, one by one, before moving on.
- __`iodide.await(dataFetch1, dataFetch2, ...)`__ maybe this is a better function - something that actually works a bit like `Promise.all`, but whose intended purpose is to wait for whatever Promises are thrown in it to resolve before popping off the cell eval queue. THIS feels like a really nice way of handling the issue, since it decouples the Promise resolution from the `fetch` operation itself, is 100% explicit, and is pretty intelligible from an API perspective.
- __load data cell__ - if we take the loading of data out of the code workflow entirely (as an option), that might solve some % of loading-data issues very nicely. This will not solve all issues, but it'll solve the analyst data exploration one.

At any rate, I hope one of these alternatives might make more sense. We can build any of these using the exact same `evalQueue` API that powers `requireExplicitContinuation` and `continue`. Those are necessities for all sorts of more nuanced and important work flows, and are great building blocks for other options.

**Comment by hamilton** [comment]
We'll be pursuing the concept behind `iodide.evalQueue.await`, which is implemented in #618.

**Comment by bcolloran** [comment]
another data point --  https://beta.observablehq.com/@mbostock/introduction-to-promises

"Observable implicitly awaits promises across cell boundaries, so you often don’t need to deal with a promise directly"

they also have a few other related functions in their stdlib

**Comment by mdboom** [comment]
From dogfooding, I think the "waiting across cell boundaries" is the right approach.  I don't think we ever want cells running in parallel -- or if we do, I think that's a 2.0 feature.

The `await` thing works (and could be the basis for that), but you have to know about it.
